File: packages/pvm/src/host-functions/accumulate/upgrade.ts
import { bytesToHex } from '@pbnj/core'
import type { HostFunctionResult } from '@pbnj/types'
import { ACCUMULATE_FUNCTIONS, RESULT_CODES } from '../../config'
import {
  type AccumulateHostFunctionContext,
  BaseAccumulateHostFunction,
} from './base'

/**
 * UPGRADE accumulation host function (Ω_U)
 *
 * Upgrades service code hash and gas limits
 *
 * Gray Paper Specification:
 * - Function ID: 19 (upgrade)
 * - Gas Cost: 10
 * - Parameters: registers[7-9] = o, g, m
 *   - o: code hash offset in memory (32 bytes)
 *   - g: new minimum accumulation gas
 *   - m: new minimum memory gas
 * - Returns: registers[7] = OK or error code
 *
 * Gray Paper Logic:
 * 1. Read code hash from memory (32 bytes)
 * 2. Update current service's code hash
 * 3. Update minimum accumulation gas
 * 4. Update minimum memory gas
 */
export class UpgradeHostFunction extends BaseAccumulateHostFunction {
  readonly functionId = ACCUMULATE_FUNCTIONS.UPGRADE
  readonly name = 'upgrade'

  execute(context: AccumulateHostFunctionContext): HostFunctionResult {
    const { registers, ram, implications } = context
    // Extract parameters from registers
    const [codeHashOffset, newMinimumAccumulationGas, newMinimumMemoryGas] =
      registers.slice(7, 10)

    // Log all input parameters
    context.log('UPGRADE host function invoked', {
      codeHashOffset: codeHashOffset.toString(),
      newMinimumAccumulationGas: newMinimumAccumulationGas.toString(),
      newMinimumMemoryGas: newMinimumMemoryGas.toString(),
      currentServiceId: implications[0].id.toString(),
    })

    // Read code hash from memory (32 bytes)
    const [codeHashData, faultAddress] = ram.readOctets(codeHashOffset, 32n)
    if (faultAddress) {
      this.setAccumulateError(registers, 'WHAT')
      return {
        resultCode: RESULT_CODES.PANIC,
      }
    }
    if (!codeHashData) {
      this.setAccumulateError(registers, 'WHAT')
      return {
        resultCode: RESULT_CODES.PANIC,
      }
    }

    // Get the current implications context
    const [imX] = implications

    // Get current service account
    const serviceAccount = imX.state.accounts.get(imX.id)
    if (!serviceAccount) {
      this.setAccumulateError(registers, 'HUH')
      return {
        resultCode: null, // continue execution
      }
    }

    // Update service account with new code hash and gas limits
    // Gray Paper: imX.self.codehash = c, imX.self.minaccgas = g, imX.self.minmemogas = m
    serviceAccount.codehash = bytesToHex(codeHashData)
    serviceAccount.minaccgas = newMinimumAccumulationGas
    serviceAccount.minmemogas = newMinimumMemoryGas

    imX.state.accounts.set(imX.id, serviceAccount)

    // Set success result
    this.setAccumulateSuccess(registers)
    return {
      resultCode: null, // continue execution
    }
  }
}

File: packages/pvm-rust/src/host_functions/accumulate/upgrade.rs
//! UPGRADE accumulation host function (Ω_U). Gray Paper: function ID 19.
//! r7 = code hash offset (o), r8 = new min accumulation gas (g), r9 = new min memory gas (m).
//! 1:1 with TS upgrade.ts: read 32-byte code hash from memory; update service account; r7 = OK or HUH/panic.

use crate::config::FUNC_UPGRADE;
use crate::host_functions::accumulate::base::{self, codes};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

const CODE_HASH_LEN: u32 = 32;

pub struct UpgradeHostFunction;

impl HostFunction for UpgradeHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_UPGRADE
    }
    fn name(&self) -> &'static str {
        "upgrade"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let code_hash_offset = context.registers[7] as u32;
        let new_min_acc_gas = context.registers[8];
        let new_min_memo_gas = context.registers[9];

        let _log_service_id = context.service_id.unwrap_or(0);
        crate::host_log!(
            "[hostfn] UPGRADE host function invoked codeHashOffset={} newMinAccGas={} newMinMemoGas={} currentServiceId={}",
            code_hash_offset, new_min_acc_gas, new_min_memo_gas, _log_service_id
        );

        let read_result = context.ram.read_octets(code_hash_offset, CODE_HASH_LEN);
        if read_result.fault_address != 0 {
            crate::host_log_error!(
                "[hostfn] upgrade PANIC: code_hash read fault (serviceId={}, offset={}, fault_address={})",
                _log_service_id, code_hash_offset, read_result.fault_address
            );
            return HostFunctionResult::panic();
        }
        let Some(data) = read_result.data else {
            crate::host_log_error!(
                "[hostfn] upgrade PANIC: code_hash read returned no data (serviceId={}, offset={})",
                _log_service_id, code_hash_offset
            );
            return HostFunctionResult::panic();
        };
        if data.len() != CODE_HASH_LEN as usize {
            crate::host_log_error!(
                "[hostfn] upgrade PANIC: code_hash length mismatch (serviceId={}, got {}, expected {})",
                _log_service_id, data.len(), CODE_HASH_LEN as usize
            );
            return HostFunctionResult::panic();
        }

        let mut codehash = [0u8; 32];
        codehash.copy_from_slice(&data);

        let service_id = match context.service_id {
            Some(id) => id as u32,
            None => {
                crate::host_log_error!(
                    "[hostfn] upgrade: service_id is None (HUH) (serviceId={})",
                    _log_service_id
                );
                base::set_accumulate_error(context.registers, codes::HUH);
                return HostFunctionResult::continue_execution();
            }
        };
        // Current service account (imX). Resolve from context.service_account or context.accounts + service_id.
        let service_account = match &mut context.service_account {
            Some(a) => a,
            None => match &mut context.accounts {
                Some(accounts) => match accounts.get_mut(&(service_id as u64)) {
                    Some(acc) => acc,
                    None => {
                        crate::host_log_error!(
                            "[hostfn] upgrade: Service account not found (serviceId={}) <- HUH",
                            _log_service_id
                        );
                        base::set_accumulate_error(context.registers, codes::HUH);
                        return HostFunctionResult::continue_execution();
                    }
                },
                None => {
                    crate::host_log_error!(
                        "[hostfn] upgrade: accounts is None (serviceId={}) <- HUH",
                        _log_service_id
                    );
                    base::set_accumulate_error(context.registers, codes::HUH);
                    return HostFunctionResult::continue_execution();
                }
            },
        };

        service_account.codehash = codehash;
        service_account.minaccgas = new_min_acc_gas;
        service_account.minmemogas = new_min_memo_gas;

        base::set_accumulate_success(context.registers, codes::OK);
        crate::host_log!("[host-calls] [{}] UPGRADE({}, {}, {}) <- OK", _log_service_id, code_hash_offset, new_min_acc_gas, new_min_memo_gas);
        HostFunctionResult::continue_execution()
    }
}
